### DISCUSSION TITLE
Can you make a pattern for loading npm packages on the client only?

### DISCUSSION THREAD

**Discussion by pvijeh** [vercel/next.js#43975] [Ideas]
### Describe the feature you'd like to request

some packages rely on DOM features such as window

### Describe the solution you'd like

these packages dont work on the server side

### Describe alternatives you've considered

we should come up with a way to load them that is safe on the server side

**Comment by Fredkiss3** [comment]
There is already such a pattern provided by the react team, to make sure a package should only be used only on the client you can install a package  named `client-only` and just import it on top of your file : 

```tsx
// src/package/my-client-only-component.tsx
import "client-only"

export const MyComponent() {
// … rest of your code

}
```

**Comment by Fredkiss3** [comment]
The same apply for server components, with the package `server-only`

**Comment by Fredkiss3** [comment]
More on the docs here : https://beta.nextjs.org/docs/rendering/server-and-client-components#keeping-server-only-code-out-of-client-components-poisoning

**Comment by SakarDev** [comment]
I have the same problem

I installed StreamSaver package which relies on **document** and since nextJS is server side by default it causes

> Server Error
> ReferenceError: document is not defined

Even installing **client-only** package did not resolve the problem.

I tried to use next/dynamic which is used to make components client side, still got the same error!!

**Comment by Fredkiss3** [comment]
@SakarDev , what `client-only` guarantee is that the component or lib can only be imported by a component marked with "use client" at the top.

client components (marked with `"use client"`)  are rendered once on the server and once on the client browser, if you want to use a lib which uses the document global, you have to call your lib in a `useEffect` if possible, if not then you can use `next/dynamic` with `ssr: false`. as of now there is a bug with dynamic ssr false, but it was fixed recently, I think the fix will be available in the next canary release (it shouldn’t take a while), to install the canary you would have to use `npm install next@canary`.

And FYI that is how it always worked in NextJs even before next 13, the difference between client components and server components is that client components run like regular next 12 components (I.e twice) and server components run only once on the server.

**Comment by SakarDev** [comment]
Thanks, i was using next/dynamic in a wrong way! It works now.
